network/p2p/middleware.go

```diff
@@ -360,8 +361,28 @@ func (m *Middleware) Unsubscribe(channel network.Channel) error {
 	return nil
 }
 
-// processMessage processes a message and eventually passes it to the overlay
-func (m *Middleware) processMessage(msg *message.Message) {
+// processMessage processes a message and a source (indicated by its PublicKey) and eventually passes it to the overlay
+func (m *Middleware) processMessage(msg *message.Message, originKey crypto.PublicKey) {
+	identities, err := m.ov.Identity()
+	if err != nil {
+		m.log.Error().Err(err).Msg("failed to retrieve identities list while delivering a message")
+		return
+	}
+
+	// check the origin of the message corresponds to the one claimed in the OriginID
+	originID := flow.HashToID(msg.OriginID)
+
+	originIdentity, found := identities[originID]
+	if !found || !originIdentity.NetworkPubKey.Equals(originKey) {
+		m.log.Warn().Msgf("message claiming to be from nodeID %v with key %x was actually signed by %x and dropped", originID, originIdentity.NetworkPubKey, originKey)
```

Comment:
Note, we may want to distinguish the error messages for the two cases above.

In particular, if `!found` is true, in the worst case you may get a segfault by trying to do `originIdentity.NetworkPubKey` on this line.

In the best case, `originIdentity.NetworkPubKey` is probably null and the error log is confusing for the reader

Reply:
I don't think the segfault is possible given `||` shortcuts: https://play.golang.org/p/g7CGIVG7dQ2
Moreover, given that we have `pubsub.WithMessageSigning(true)` and `pubsub.WithStrictSignatureVerification(true)`, I would note a messsage for which `found != true` would likely not arrive here but in theory, agreed.

Reply:
I meant that even in the log message itself, you are calling `originIdentity.NetworkPubkey`